stlab/concurrency/router.hpp

```diff
@@ -0,0 +1,137 @@
+        virtual stlab::receiver<T> get_default_route() = 0;
+        virtual void set_ready() = 0;
+        virtual void route(T t) = 0;
+        virtual ~concept_t() = default;
+    };
+
+    template <class E, class F>
+    struct model : concept_t {
+        model(E executor, F router_func)
+        : _executor{std::move(executor)}
+        , _router_func{std::move(router_func)}
+        , _sender_receiver{stlab::channel<T>(_executor)}
+        {}
+
+        void set_ready() override {
+            std::sort(std::begin(_routes), std::end(_routes), [](const route_pair& a, const route_pair& b){
+                return a.first < b.first;
+            });
+            for (auto& pair : _routes)
+                pair.second.second.set_ready();
+            _sender_receiver.second.set_ready();
+        }
+
+        stlab::receiver<T> get_default_route() override { return _sender_receiver.second; }
+
+        stlab::receiver<T> get_route(K key) override {
+            auto find_it = std::find_if(begin(_routes), std::end(_routes), [&](const route_pair& pair){
+                return pair.first == key;
+            });
+            if (find_it != std::end(_routes)) return find_it->second.second;
+            _routes.push_back(std::make_pair(std::move(key), stlab::channel<T>(_executor)));
+            return _routes.back().second.second;
+        }
+
+        template<class C, typename std::enable_if_t<std::is_same<typename C::value_type, std::pair<K, bool>>::value, int> = 0>
+        void route_keys(const C& key_set, T t) {
+            bool did_route = false;
+            auto find_it = std::begin(_routes);
+            for (const auto& key : key_set) {
+                if (!key.second) continue;
+                find_it = std::lower_bound(find_it, std::end(_routes), key, [](const route_pair& a, const std::pair<K, bool>& k){
+                    return a.first < k.first;
+                });
+                if (find_it == std::end(_routes)) break;
+                if (find_it->first != key.first) continue;
+                find_it->second.first(t);
+                did_route = true;
+            }
+            if (!did_route) return _sender_receiver.first(std::move(t));
+        }
+
+        template<class C, typename std::enable_if_t<std::is_same<typename C::value_type, K>::value, int> = 0>
+        void route_keys(const C& key_set, T t) {
+            auto find_it = std::begin(_routes);
+            for (const auto& key : key_set) {
+                find_it = std::lower_bound(find_it, std::end(_routes), key, [](const route_pair& a, const K& k){
+                    return a.first < k;
+                });
+                if (find_it == std::end(_routes)) return;
+                if (find_it->first != key) continue;
+                find_it->second.first(t);
+            }
+        }
+
+        void route(T t) override {
+            if (std::empty(_routes)) return _sender_receiver.first(std::move(t));
```

Comment:
Why `return` if this is a void function?

Reply:
Return early pattern. It just so happens that a `sender<>` function operator is a void function so I can put it all on one line. The equivalent would be:
```c++
if (std::empty(_routes)) {
  _sender_receiver.first(std::move(t));
  return;
}
```

Reply:
I removed the default route because it was not even necessary. If they want a default route they can make a normal route that they conditionally use if no other case is matched.